docs/src/design/proposals/Registration.md

```diff
@@ -0,0 +1,518 @@
+
+```
+function NavigationPlugin(core) {
+    var navigateActionProvider = new Provider(function () {
+        return new NavigateAction();
+    });
+
+    // Expose to permit extension
+    this.navigateActionProvider = navigateActionProvider;
+
+    Plugin.call(this, function () {
+        core.actionRegistry.register(function () {
+            return navigateActionProvider.get();
+        });
+    });
+}
+```
+
+And other plugins could then extend the "navigate" action directly:
+
+```
+function EditPlugin(navigationPlugin) {
+    Plugin.call(this, function () {
+        navigationPlugin.navigateActionProvider
+            .decorate(function (navigateAction) {
+                return new DecoratedNavigationAction(navigateAction);
+            });
+    });
+}
+```
+
+# Legacy Support
+
+The proposed API does not achieve parity with the existing Framework
+Layer. In particular:
+
+1. There is no Angular integration.
+2. Transitional support for legacy bundles is not specified.
+3. Script loading is not handled.
+
+Propose to resolve these by:
+
+1. Adding an Angular integration plugin that uses registries for
+   controllers/directives/etc. and maps these to appropriate configuration
+   of an Angular application.
+2. Adding a legacy support plugin that maps (and transforms, as necessary)
+   extensions exposed by legacy bundles to appropriate registries and/or
+   service providers using the new Registration API.
+3. Not handling script loading; the Registration API expects objects,
+   so how the scripts necessary to instantiate those objects (and how they
+   are instantiated) is left to the application developer.
+
+# Evaluation
+
+[Identified problems](#problems-to-address) are addressed by this solution:
+
+1. Dependencies between plugins can be made explicit; a `Plugin` may
+   impose dependencies on other specific `Plugin` subclasses as constructor
+   arguments, disambiguated with JSDoc. The software does not take part
+   in dependency management among plugins; rather, this responsibility is
+   plainly communicated to developers.
+2. Extension points are made explicit; `Provider` instances must be
+   reachable for plugins to configure, and may be made available as
+   public fields of `Plugin`s. Their types can be clearly documented,
+   usages and interactions can be followed with standard developer tools
+   (e.g. breakpoints), and so on.
+3. Reuse of classes between plugins is neither facilitated nor impeded
+   by the registration API. If, however, plugins are written following the
+   "expose classes in namespaces" approach, then it is trivially to
+   expose additional classes in these same namespaces.
+
+This solution offers further benefits:
+
+* The `Provider` API is robust enough to support the various existing
+  extensions of Open MCT, but its usage is opt-in; plugins are free
+  to expose other (potentially wildly different) means of extension.
+  Usage of `Provider`s is _encouraged_ to promote ubiquitous
```

Comment:
I would say "Usage of provider is encouraged when defining new extension points inside of the core platform."  Don't expect many plugin developers to care about this.  Only platform developers would care.